torch/distributed/fsdp/flat_param.py

```diff
@@ -281,7 +338,7 @@ def _init_flat_param(
         self.flat_param = FlatParamHandle.flatten_params(
```

Comment:
where do we set requires_grad for the shared paras if section? Maybe i missed it with the nested loops. If it is only for the non shared params, should we move it closer to where it is assigned?

Reply:
Shared parameters in this context refers to two separate Python variables that reference the same `nn.Parameter`. Practically, this can happen in transformer architectures that share the embedding and unembedding weight matrices.

Because shared parameters reference the same `nn.Parameter`, they are guaranteed to have the same `requires_grad`, which is why we do not need to consider `requires_grad` in the shared parameters `if` branch.

Maybe the currently `requires_grad = None` and then `requires_grad = param.requires_grad` every iteration is confusing. This is just to have `requires_grad` be the uniform `requires_grad` value of all parameters by the end of the loop. After the first `requires_grad = param.requires_grad`, the remaining assignments are all vacuous. 

I am not sure if we can move the `requires_grad` assignment any closer to the usage in `FlatParamHandle.flatten_params()` (beyond within the same `else` branch) since (1) we need to get the `requires_grad` from inside the loop and (2) we need to finish the loop.